Introducing statically-linked plugins

evan2645 opened the pull request:
This commit teaches the common catalog to handle "builtin" plugins. Builtin plugins are, for all intents and purposes, identical to plugins run out-of-process: they conform to the same interfaces, are configured the same way, and are consumed from the catalog in the same way. For the consumer, the experience is identical. For users, they will no longer need to worry about managing multiple processes. As more plugin options are introduced, the user can choose which to enable by simply configuring them as enabled, similar to the way they would do with an out-of-process plugin.

Agent and server specific catalogs will understand which plugins should be linked in, generating references to them and passing them into the common catalog. This commit updates the server catalog to statically link all existing checked-in plugins it relies upon. A subsequent PR will introduce similar changes in the agent's catalog.

I will note that with statically-linked plugins, a number of previously-held conventions no longer make sense... for instance, we could probably do without the `enabled` option, and perhaps the format of the configuration file could benefit from a light refactor. Additionally, the catalog has undergone a number of changes in recent months (including this one), and could itself benefit from a refactor to clean it up a bit. Finally, the testing methodologies used for the various plugins are a bit out-of-place, and I have had to make some minor updates to plugin instantiation and testing code in order to get everything working smoothly. I have stopped short of actually attempting any of these refactors in order to keep this change well scoped.

The top-level plugin directory doesn't make much sense for builtins, so I have moved them to a new subdirectory in the server package. I have also done some light renaming in order for things to be a little more idiomatic.

I expect at least a couple refactors to follow this. The goal of this change was to introduce statically-linked plugins without boiling the oceans, and a lot of the spots needing refactor (like CA plugin tests) were in need of refactors prior to this change anyways.

Now, let's get that AWS plugin set checked in :).

@drrt for build and packaging considerations
@walmav for functional testing considerations
@ajessup for user experience considerations

walmav commented on pkg/common/catalog/catalog.go:
```diff
@@ -177,7 +184,13 @@ func (c *catalog) loadConfigFromHclConfig(hclPluginConfig HclPluginConfig) error
 	for _, p := range c.plugins {
 		if !p.Config.Enabled {
-			c.l.Debugf("%s plugin %s is disabled and will not be started", p.Config.PluginType, p.Config.PluginName)
+			c.l.Debugf("%s plugin %s is disabled", p.Config.PluginType, p.Config.PluginName)
```
consider using k:v style logging:
c.l.Debugf("plugin_type:%s plugin plugin_name:%s is disabled", p.Config.PluginType, p.Config.PluginName)

amartinezfayo replied:
Since there are multiple messages "%s plugin %s is disabled" I found useful to have an indication of the context, when looking at the output log. That's why it says :  "%s plugin %s is disabled and will not be started".
E.g., we have other places where it says.: "%s plugin %s is disabled and will not be configured" and "%s plugin %s is disabled and will not be categorized".

@evan2645  Does this make sense to you?

evan2645 replied:
@amartinezfayo yes, reverted.

@walmav agreed these should be fields (logrus has first-class support for field logging). I am hesitant to change them now though since there are lots of other places in this catalog that don't use fields and the update would make these bits inconsistent. I mentioned that catalog could probably benefit from a refactor in the coming months, we should make field logging an explicit goal of that refactor, including consistent use of field names etc. I assume this is non-blocking since you approved? Will change it over if you feel strongly

walmav commented on pkg/common/catalog/catalog.go:
```diff
@@ -274,3 +276,38 @@ func (c *catalog) newPluginConfig(p *ManagedPlugin) (*goplugin.ClientConfig, err
+
+func (c *catalog) secureConfig(p *ManagedPlugin) (*goplugin.SecureConfig, error) {
+	if p.Config.PluginChecksum == "" {
+		c.l.Warnf("%s plugin %s not using secure config", p.Config.PluginType, p.Config.PluginName)
```
same here consider using k:v style logging

drrt commented:
[x] Docs update?

evan2645 commented:
Thanks for keeping me honest @drrt :)

I have updated the README accordingly, and added a table of builtins. The exercise also led me to realize that the name `memory` for UpstreamCA plugin is silly, so I have changed it to `disk`.
